### PR TITLE
Fix BackButton handling input before all other UI elements

### LINKED ISSUE #5177: SearchTextBox no longer clears text when pressing Escape
**Describe the bug:** The expected behaviour, when pressing `Esc`, is that the `SearchTextBox` on the carousel clears the text it contains. After that, the next `Esc` key pressed should return to the Menu.

**Screenshots or videos showing encountered issue:** https://streamable.com/116sk

**osu!lazer version:** 2019.628.0

### PR DESCRIPTION
**Pull request by nyquillerium**
Fixes https://github.com/ppy/osu/issues/5177

Attempt 2, now with 100% more automated testing.

This approach just rebuilds the `KeyBindingInputQueue` with the back button at the end.

### CODE REVIEW

File: osu.Game.Tests/Visual/Menus/TestSceneExitingScreens.cs [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+                    case null:
+                    case Intro _:
+                    case Disclaimer _:
+                        return false;
+
+                    case MainMenu _:
+                        return true;
+
+                    default:
+                        current.Exit();
```
Comment (iiSaLMaN) [inline osu.Game.Tests/Visual/Menus/TestSceneExitingScreens.cs]:
```suggestion
                        current.Exit();
                    case null:
                    case Intro _:
                    case Disclaimer _:
```

shortens the switches

File: osu.Game.Tests/Visual/Menus/TestSceneExitingScreens.cs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+                        return false;
+                }
+            });
+        }
+
+        [Test]
+        public void TestExitingSongSelectWithEscape()
+        {
+            TestSongSelect songSelect = null;
```
Comment (iiSaLMaN) [inline osu.Game.Tests/Visual/Menus/TestSceneExitingScreens.cs]: redundant assign, same for the other tests

File: osu.Game.Tests/Visual/Menus/TestSceneExitingScreens.cs [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+            AddStep("Push songselect", () => osuGame.ScreenStack.Push(songSelect = new TestSongSelect()));
+            AddUntilStep("Wait for song select", () => songSelect.IsCurrentScreen());
```
Comment (iiSaLMaN) [inline osu.Game.Tests/Visual/Menus/TestSceneExitingScreens.cs]: move to a function, probably `pushScreenStep(new TestSongSelect())`

File: osu.Game.Tests/Visual/Menus/TestSceneExitingScreens.cs [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+                }
+            });
+        }
+
+        [Test]
+        public void TestExitingSongSelectWithEscape()
+        {
+            TestSongSelect songSelect = null;
+
+            AddStep("Push songselect", () => osuGame.ScreenStack.Push(songSelect = new TestSongSelect()));
```
Comment (iiSaLMaN) [inline osu.Game.Tests/Visual/Menus/TestSceneExitingScreens.cs]: inconsistent name, add a space between `song` `select` *might be unnecessary*

**Review by frenzibyte — CHANGES_REQUESTED** [review]

File: osu.Game.Tests/Visual/Menus/TestSceneExitingScreens.cs (hunk 1 above)
Comment (nyquillerium) [reply osu.Game.Tests/Visual/Menus/TestSceneExitingScreens.cs]: I don't think this works unless I'm mistaken, but I can probably simplify this switch another way.

File: osu.Game.Tests/Visual/Menus/TestSceneExitingScreens.cs (hunk 2 above)
Comment (nyquillerium) [reply osu.Game.Tests/Visual/Menus/TestSceneExitingScreens.cs]: required because test steps may be manually ran out of order, which results in a warning that this value may be used without being assigned.

**Comment by smoogipoo** [comment]
On discussion with @peppy, it might be cleaner to do this via the action receptor model.

```yaml
Children:
    backButtonReceptor
    screenStack
    backButton(backButtonReceptor)

class BackButton:
    ctor(BackButtonReceptor receptor = null):
        receptor.OnPressed += ...
        receptor.OnReleased += ...
```

File: osu.Game/OsuGame.cs [hunk 5]
```diff
@@ -81,10 +81,13 @@ public class OsuGame : OsuGameBase, IKeyBindingHandler<GlobalAction>
 
         public readonly Bindable<OverlayActivation> OverlayActivationMode = new Bindable<OverlayActivation>();
 
-        private OsuScreenStack screenStack;
+        protected OsuScreenStack ScreenStack;
+        protected BackButton BackButton;
+
+        private BackButton.BackButtonReceptor backButtonReceptor;
```
Comment (peppy) [inline osu.Game/OsuGame.cs]: is there a reason this isn't a local?

File: osu.Game/Graphics/UserInterface/BackButton.cs [hunk 6]
```diff
@@ -61,5 +50,24 @@ protected override void PopOut()
             button.MoveToX(-TwoLayerButton.SIZE_EXTENDED.X / 2, 400, Easing.OutQuint);
             button.FadeOut(400, Easing.OutQuint);
         }
+
+        public class BackButtonReceptor : Drawable, IKeyBindingHandler<GlobalAction>
```
Comment (peppy) [inline osu.Game/Graphics/UserInterface/BackButton.cs]: Since this is nested, just call it `Receptor`

File: osu.Game/Graphics/UserInterface/BackButton.cs [hunk 7]
```diff
@@ -10,14 +10,16 @@
...
+    public class BackButton : VisibilityContainer
     {
         public Action Action;
 
         private readonly TwoLayerButton button;
 
-        public BackButton()
+        public BackButton(BackButtonReceptor receptor)
         {
+            receptor.OnBackPressed += () => Action.Invoke();
```
Comment (peppy) [inline osu.Game/Graphics/UserInterface/BackButton.cs]:
```suggestion
            receptor.OnBackPressed = () => Action.Invoke();
```

File: osu.Game.Tests/Visual/Menus/TestSceneExitingScreens.cs [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+                },
+                osuGame
+            };
+        }
+
+        [SetUpSteps]
+        public void SetUpSteps()
+        {
+            AddUntilStep("wait for load", () => osuGame.IsLoaded);
+            AddUntilStep("wait for main menu", () =>
```
Comment (peppy) [inline osu.Game.Tests/Visual/Menus/TestSceneExitingScreens.cs]: this isn't "waiting", this is "exiting"

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game.Tests/Visual/Menus/TestSceneExitingScreens.cs [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+                },
+                osuGame
+            };
+        }
+
+        [SetUpSteps]
+        public void SetUpSteps()
+        {
+            AddUntilStep("wait for load", () => osuGame.IsLoaded);
+            AddUntilStep("exit to main menu", () =>
```
Comment (peppy) [inline osu.Game.Tests/Visual/Menus/TestSceneExitingScreens.cs]: what's the purpose of this step? how would we be anywhere else?

Comment (nyquillerium) [reply osu.Game.Tests/Visual/Menus/TestSceneExitingScreens.cs]: We could potentially still be in the intro or disclaimer, and this makes sure you can reset a test in visual test browser even if you stop in the middle of one.

Comment (peppy) [reply osu.Game.Tests/Visual/Menus/TestSceneExitingScreens.cs]: that's not how it should be done. "resetting" should be making a new game. you're just asking for this to fail in random ways.

Comment (nyquillerium) [reply osu.Game.Tests/Visual/Menus/TestSceneExitingScreens.cs]: I guess we should create a game instance that skips directly to the main menu on run to avoid this step entirely.

Comment (peppy) [reply osu.Game.Tests/Visual/Menus/TestSceneExitingScreens.cs]: the step is fine. it shouldn't be calling exit

**Comment by smoogipoo** [comment]
Exiting from overlays is not tested.

File: osu.Game/Screens/Multi/Multiplayer.cs [hunk 10]
```diff
@@ -166,7 +168,7 @@ public void Start(Func<Player> player)
 
         public void APIStateChanged(IAPIProvider api, APIState state)
         {
-            if (state != APIState.Online)
+            if (RequireOnline && state != APIState.Online)
```
Comment (peppy) [inline osu.Game/Screens/Multi/Multiplayer.cs]: can't this instead be done via the dummy api access?

File: osu.Game/Overlays/MusicController.cs [hunk 11]
```diff
@@ -316,6 +316,10 @@ private void prev()
 
         private void next(bool instant = false)
         {
+            // beatmapSets doesn't get populated until loading has completed.
```
Comment (peppy) [inline osu.Game/Overlays/MusicController.cs]: why is this here? if this is causing a crash or issue it should 100% be its own PR with own tests.

Comment (nyquillerium) [reply osu.Game/Overlays/MusicController.cs]: this is causing a crash only on headless runs when going through the load process (since loading the `MusicController` can take a while). Can split this out.

Comment (peppy) [reply osu.Game/Overlays/MusicController.cs]: very highly doubt it's headless only. nothing stops a desktop run from also taking equally as long.

Comment (nyquillerium) [reply osu.Game/Overlays/MusicController.cs]: Turns out this was unnecessary and was an issue in my test

File: osu.Game.Tests/Visual/Menus/TestSceneScreenNavigation.cs [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+        public void TestExitMultiWithBackButton()
+        {
+            pushAndConfirm(() => new Screens.Multi.Multiplayer(), "multiplayer");
+            exitViaBackButtonAndConfirm();
+        }
+
+        [Test]
+        public void TestOpenOptionsAndExitWithEscape()
+        {
+            AddUntilStep("Wait for options to load", () => osuGame.Settings.IsLoaded);
```
Comment (peppy) [inline osu.Game.Tests/Visual/Menus/TestSceneScreenNavigation.cs]: this shouldn't be required.

File: osu.Game.Tests/Visual/Menus/TestSceneScreenNavigation.cs [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+            AddStep("Enter menu", () => pressAndRelease(Key.Enter));
+            AddStep("Move mouse to options overlay", () => InputManager.MoveMouseTo(optionsButtonPosition));
+            AddStep("Click options overlay", () => InputManager.Click(MouseButton.Left));
+            AddAssert("Options overlay was opened", () => osuGame.Settings.State.Value == Visibility.Visible);
+            AddStep("Hide options overlay using escape", () => pressAndRelease(Key.Escape));
+            AddAssert("Options overlay was closed", () => osuGame.Settings.State.Value == Visibility.Hidden);
+        }
+
+        private void pushAndConfirm(Func<Screen> newScreen, string screenName)
```
Comment (peppy) [inline osu.Game.Tests/Visual/Menus/TestSceneScreenNavigation.cs]: you don't need the `string` argument

File: osu.Game.Tests/Visual/Menus/TestSceneScreenNavigation.cs [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+                }
+
+                osuGame = new TestOsuGame();
+                osuGame.SetHost(gameHost);
+
+                Add(osuGame);
+            });
+            AddUntilStep("Wait for load", () => osuGame.IsLoaded);
+            AddUntilStep("Wait for intro", () => osuGame.ScreenStack.CurrentScreen is IntroScreen);
+            AddUntilStep("Wait for main menu", () => osuGame.ScreenStack.CurrentScreen is MainMenu menu && menu.IsLoaded);
```
Comment (peppy) [inline osu.Game.Tests/Visual/Menus/TestSceneScreenNavigation.cs]: this line is in 3 places

File: osu.Game.Tests/Visual/Menus/TestSceneScreenNavigation.cs [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+using osuTK.Input;
+
+namespace osu.Game.Tests.Visual.Menus
+{
+    public class TestSceneScreenNavigation : ManualInputManagerTestScene
+    {
+        private GameHost gameHost;
+        private TestOsuGame osuGame;
+
+        private Vector2 backButtonPosition => osuGame.ToScreenSpace(new Vector2(25, osuGame.LayoutRectangle.Bottom - 25));
```
Comment (peppy) [inline osu.Game.Tests/Visual/Menus/TestSceneScreenNavigation.cs]: you can make 25 a constant `click_padding` or something.

**Review by peppy — DISMISSED** [review]

File: osu.Game.Tests/Visual/Menus/TestSceneScreenNavigation.cs (hunk 12 above)
Comment (nyquillerium) [reply osu.Game.Tests/Visual/Menus/TestSceneScreenNavigation.cs]: This is required because this actually takes quite a long time in headless, and trying to toggle the options overlay off while its not loaded with escape will not work since the `FocusedTextBox` has not taken focus yet.

File: osu.Game/OsuGame.cs [hunk 16]
```diff
@@ -96,14 +100,14 @@ public class OsuGame : OsuGameBase, IKeyBindingHandler<GlobalAction>
...
 
-        private SettingsPanel settings;
-
         private readonly List<OverlayContainer> overlays = new List<OverlayContainer>();
 
         private readonly List<OverlayContainer> toolbarElements = new List<OverlayContainer>();
 
         private readonly List<OverlayContainer> visibleBlockingOverlays = new List<OverlayContainer>();
 
+        protected virtual Loader CreateLoader() => new Loader();
```
Comment (smoogipoo) [inline osu.Game/OsuGame.cs]: Method should not be above ctor

File: osu.Game/Graphics/UserInterface/BackButton.cs [hunk 17]
```diff
@@ -61,5 +50,24 @@ protected override void PopOut()
...
+        public class Receptor : Drawable, IKeyBindingHandler<GlobalAction>
+        {
+            public Action OnBackPressed;
+
+            public bool OnPressed(GlobalAction action)
+            {
+                switch (action)
+                {
+                    case GlobalAction.Back:
+                        OnBackPressed.Invoke();
```
Comment (smoogipoo) [inline osu.Game/Graphics/UserInterface/BackButton.cs]:
```suggestion
                        OnBackPressed?.Invoke();
```

File: osu.Game/Graphics/UserInterface/BackButton.cs [hunk 18]
```diff
@@ -10,14 +10,16 @@
...
+    public class BackButton : VisibilityContainer
     {
         public Action Action;
 
         private readonly TwoLayerButton button;
 
-        public BackButton()
+        public BackButton(Receptor receptor)
         {
+            receptor.OnBackPressed = () => Action.Invoke();
```
Comment (smoogipoo) [inline osu.Game/Graphics/UserInterface/BackButton.cs]:
```suggestion
            receptor.OnBackPressed = () => Action?.Invoke();
```

**Review by smoogipoo — CHANGES_REQUESTED** [review]

**Review by smoogipoo — APPROVED** [review]
